pysisyphus/calculators/CFOUR.py

```diff
@@ -0,0 +1,191 @@
+import re
+import shutil
+import textwrap
+
+import numpy as np
+
+from pysisyphus.calculators.Calculator import Calculator
+
+class CFOUR(Calculator):
+
+    conf_key = "cfour"
+
+    def __init__(
+            self,
+            cfour_input,
+            keep_molden=True,
+            initden_file=None,
+            **kwargs,
+    ):
+        """CFOUR calculator.
+
+        Wrapper handling CFOUR ground state energy and gradient calculations.
+
+        Parameters
+        ----------
+        cfour_input : dict
+            CFOUR keywords and values. Note: "on" must be encapsulated in quotes to avoid being translated to True by YAML.
+        keep_molden : bool, optional
+            Whether or not to keep ground state SCF orbitals for each geometry step.
+        initden_file: str, optional
+            Path to an input initden file for use as a guess SCF density.
+        """
+        super().__init__(**kwargs)
+
+        self.cfour_input = cfour_input
+
+        self.inp_fn = 'ZMAT'
+        self.out_fn = 'out.log'
+        self.gradient_fn = 'GRD'
+        self.to_keep = ("out.log", "density:__den.dat")
+        self.initden = None
+        if keep_molden:
+            self.to_keep = self.to_keep + ("MOLDEN*",)
+        if initden_file:
+            self.initden = initden_file
+
+        self.base_cmd = self.get_cmd("cmd")
+        self.parser_funcs = {
+            "energy": self.parse_energy,
+            "grad": self.parse_gradient,
+        }
+
+        self.float_regex = r"([-]?\d+\.\d+)"  ## CFOUR doesn't use scientific notation for final energy or gradient.
+
+        ## Convert pysisyphus keywords to CFOUR keywords
+        ## Explicitly not adding support for managing CFOUR parallelism - 
+        ## user-configured cmd script in .pysisyphusrc should set either OMP_NUM_THREADS 
+        ## or the environment variable they defined at compile-time for MPI parallelism
+        if 'mem' in kwargs:
+            cfour_input['MEM_SIZE'] = kwargs['mem']
+            cfour_input['MEM_UNIT'] = 'MB'
+        if 'charge' in kwargs:
+            cfour_input['CHARGE'] = kwargs['charge']
+        if 'mult' in kwargs:
+            cfour_input['MULT'] = kwargs['mult']
+
+    def prepare(self, inp):
+        path = super().prepare(inp)
+        if self.initden:
```

Comment:
I guess this is an inital guess/converged MOs from a previous calculation?

Reply:
Yes, but unfortunately it's a density rather than MO coefficients. CFOUR has two ways of feeding in guesses for the SCF, both only mildly documented:

1. Feed in a file named OLDMOS which contains MO coefficients
2. Feed in a file called initden.dat which contains an initial guess electronic density

There's a case in which initden will work and OLDMOS won't. I think it has something to do with symmetry/occupation number, but my memory is not good enough to be confident on that. I chose the more robust option in this case, but it could later be extended to also allow for OLDMOS.

Reply:
This is fine; I guess it doesn't matter that it is the density, because this is what a QC code constructs from a given set of MOs...